.github/workflows/test.yml

```diff
@@ -1,12 +1,86 @@
 name: Tests
-on: push
+on: [pull_request, push]
 jobs:
-  tests:
+  linting:
+    name: Linting
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v2
+
+      - name: Setup Python 3.8
+        uses: actions/setup-python@v1
+        with:
+          python-version: "3.8"
+          architecture: x64
+
+      - uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-linting
+          restore-keys: ${{ runner.os }}-linting
+
+      - run: pip install bandit black codespell flake8 isort mypy pyupgrade safety
+      - run: bandit -r . || true
+      - run: black --check . || true
+      - run: codespell --ignore-words-list="mape" --quiet-level=2  # --skip=""
+      - run: flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+      - run: isort --check-only --profile black . || true
+      - run: mypy --ignore-missing-imports . || true
+      - run: shopt -s globstar && pyupgrade --py36-plus **/*.py || true
+      - run: safety check
+  test:
+    name: ubuntu-latest - Python 3.8
+    needs: linting
     runs-on: ubuntu-latest
```

Comment:
Do you want to `fast-fail: false` to see the results of all failing tests?

Reply:
I went back and forth on this. I want to try and limit the number of minutes as much as possible. I feel like failing fast might be the best way to do that. I feel like this project will attract a lot of new contributors who will push lots of failing builds.